### ISSUE TITLE
What is the Magit specific rewrite functionality good for?

### ISSUE THREAD

**Issue by tarsius** [magit/magit#966]
TL;DR: _rewrite_ contains some goodies that should be generalized but on its own it is not really useful and actually harmful.  I want to generalize the good parts and remove the bad ones.  This will take some time and shouldn't all be done now (before making a release).  But I want to remove `magit-insert-pending-changes` now and without a replacement.

---

What problem does the _rewrite_ functionality solve that neither `git cherry-pick`, `git rebase  --interactive`, nor `git merge --rebase` deal with equally well (or better)?

I get the impression that _rewrite_ is a bastard to `git-cherry-pick` and `git-rebase --interactive`.  In an interesting twist _rewrite_ from a user perspective works more like _rebase NOT-interactive_ (replay _these_ commits over there) than _cherry-pick_ (get those commits and apply them _here_), but at the same time it is actually implemented using _cherry-pick_ and also invoked like that (with the "destination", not the "source", being checked out).  There might be some benefit in this, but I find it to be rather confusing (being comfortable with using vanilla _rebase_ and _cherry-pick_).

Regardless, if we come to the conclusion that this is a good thing, we could implement this by using `git cherry-pick' on a range and let git keep track of what remains to be done, instead of what we do now: build the list ourselves and the cherry-pick each commit individually.  We don't need to come up with our own temporary files inside .git and then effectively implement a incomplete variant of _cherry-pick_ that has some of the properties of _rebase_.  Since _interactive rebase_ is superior to _cherry-pick_ we should also consider using the former instead.
- For _rewrite_ Magit has two convenience features that are missing for _rebase_ and _cherry-pick_: In the status buffer a list of not-yet-applied commits and `git diff -R HEAD_BEFORE_FIRST_CHERRY_PICK` are displayed.  
  
  Such lists could be (and will be) trivially implemented for _rebase_ and _cherry-pick_, so that isn't an argument in favor of _rewrite_.
  
  I am not sure that the diff is actually useful.  I find the reversal of the patch especially confusing.  Even if this diff is somehow useful it should _not_ be displayed in the status buffer.  There are other situations in which one would want to see the difference between two obvious states (e.g. `current-branch..upstream-branch`) but we don't display the diff in the status buffer just because.  We have a buffer dedicated to this: `*magit-diff*`.  Why should we make an exception for _rewrite_?
  
  I think we should remove this "status section" now (even before integrating _rewrite_ better with vanilla _cherry-pick_).  If (and only if) someone complains about its removal we could then provide a substitute: extend (or wrap in a new command) `magit-diff` so that it automatically uses the appropriate range (and `-R`) when invoked during a _rewrite_.
  
  (
  
  This "diff that automatically knows what I want to compare, but only in this one particular case" feature is not something I would implement in the first place, but now I might, or might not, be forced to do so.  The only reason why I would write it is that it would be a substitute for a feature that was removed.  I would not have implemented that other feature either but since some users might have come to expect it, I would have to implement a substitute, so that nobody can say that I am removing features just because.
  
  And by the way I think the _rewrite_ functionality is actually harmful.  Compared to _cherry-pick_ and especially _rebase_, it produces inferior output.  I have actually been given the result of somebody else using _rewrite_ and had to redo the work with _rebase_ because the _rewrite_ output just wasn't usable (e.g. lost merges).  And I have seen other instances of _rewrite_ gone wrong (non-interactive rebase should have been used instead). 
  
  So I am not (suggesting to) removing a feature because _I personally don't use it_, I remove it because _it is misguided and even harmful_.  And as this post proves, I don't just jump to that conclusion, I reason about why that is so and whether that is due to some missing features and/or bugs - or due to a fundamental problem with the underlying idea.
  
  )
- The biggest difference (from a user perspective) between _rebase_, _cherry-pick_, and _rewrite_ is how the ordering and stepping is done.
  
  **Interactive rebase**: (1) user selects range (2) user optionally reorders and sets breakpoints (3) git runs until conflict or breakpoint (4) user resolves conflict and/or makes other changes (5) user tells git to continue.  3,4,5 might be repeated several times.  More importantly though it is possible at 4 to go back to 2 (for what hasn't been applied yet, obviously).  _Interactive rebase_ is far more powerful than the alternatives, I have even left out other things it can do.
  
  **Cherry-Pick**: (1) user selects a range (3) git runs until conflict (4) user resolves conflict (5) user tells git to continue.  Again 3,4,5 might be repeated several times.  There is no 2 (reorder commits and set breakpoint).  There is no way to teach _cherry-pick_ about breakpoints.  But it would be possible to allow the reordering of commits, but only after the first conflict has happened, so implementing this is rather pointless.
  
  **Rewrite**: (1) user selects a range (2') the commits appear in the status buffer in a fixed non-editable order (5') user "continues" by manually applying individual commits using `magit-cherry-pick-item` on a single commit from the list. _Rewrite_ does not help much in doing so: (a) after a commit was applied the user manually has to also use `magit-rewrite-set-used` to make it disappear from the list, and (b) the order in the list is fixed, if the user want a different order then he has to remember that order himself. There also is a command to apply all remaining commits at once, but that is done by repeatedly using `git cherry-pick` on a single commit.

---

In my opinion _rewrite_ effectively does the following:
- Lure users into believing it is a _cherry-pick_ or _interactive rebase_ on steroids, while in reality it is a crippled version of _cherry-pick_.
- Provide a list of commits "yet to be applied" that appears in a prominent place and is not as risk of being lost.
  - Having a list of not-yet-applied commits in the status buffer can also be done for _cherry_ and _interactive rebase_, and I will do just that.
  - Having such lists inside the status buffer is both _convenient_ as well as (to some extend) _necessary_.
  - It is _necessary_ because the alternative would be to display the list in the `*magit-log*` buffer, but a list displayed there is at risk of being replaced by another list at any time.
  - Whether reusing `*magit-log*` for all logs, effectively only ever displaying the "latest log" is a good thing, is another question entirely.  (E.g. we might want to _consider_ to at least using a separate buffer for each repository.)
  - However there is an easy way to display multiple logs (or diffs etc.) at the same time: just rename the buffer.
  - Renaming the buffer should be made more convenient by providing `magit-rename-buffer` which would automatically suggest a name like e.g.: `*magit-log: HEAD~5..master @this-repository*`.

**generalize the good parts**:

A general purpose "list of random commits" in the status buffer is actually useful.  But it should be decoupled from _rewrite_ logic and be generalized.  The list should be a sort of "registry" that can be used to remember all kinds of commits.  It _could_ be used to hold a list of commits to-be-cherry-picked, but it could also be used to hold a list of commits that introduce bugs that still need to be fixed, etc.  Marking a commit would cause it to appear in this list; ranges could also be marked by marking the region; this does not need to conflict with the concept of having _the, one and only_ marked commit; and this is in addition to the _rebase_ and _real cherry-pick_ commit lists.

**remove the bad/unneeded parts**:

After the "outstanding cherries" list has been converted to a general-purpose registry, there isn't much left in _rewrite_, and it should be removed.  Provided the above has been implemented that would mean: (a) remove `magit-rewrite-start`, `magit-mark` on a region would server the same purpose (b) remove `magit-finish`, `magit-cherry-pick` on the heading of the registry list would serve the same purpose (c) remove `magit-rewrite-abort`, `magit-registry-empty` would empty the registry but _not_ hard reset to some earlier state.
